### PR TITLE
Fix ScalaMock for SJS 1.x

### PR DESCRIPTION
**Pull request by ddworak**
# Pull Request Checklist

* [x] I agree to licence my contributions under the [MIT licence](https://github.com/paulbutcher/ScalaMock/blob/master/LICENCE)
* [x] I have added copyright headers to new files
* [x] I have added tests for any changed functionality

## Purpose

Hi! During the migration to SJS 1.x in one of our projects, I've noticed practically none of our mocks made with ScalaMock work, failing with a runtime error. You can replicate the issue on my branch at the point where I enabled JS testing (def240d686c46e86479f4386d3e732d9c87ad9ae) by running `scalamockJS/test`:

```
java.lang.IllegalArgumentException: Property 'mock$oneParamMethod$0$1' is not defined in 
'org.scalamock.test.specs2.FixtureContextTest$TestSetup$$anon$1@b'. 
Available properties: 
Lorg_scalamock_test_specs2_FixtureContextTest$TestSetup$$anon$1__f_mock$special$mockName,
Lorg_scalamock_test_specs2_FixtureContextTest$TestSetup$$anon$1__f_mock$polymorphicMethod$0,
Lorg_scalamock_test_specs2_FixtureContextTest$TestSetup$$anon$1__f_mock$oneParamMethod$0,
Lorg_scalamock_test_specs2_FixtureContextTest$TestSetup$$anon$1__f_mock$noParamMethod$0,
Lorg_scalamock_test_specs2_FixtureContextTest$TestSetup$$anon$1__f_bitmap$init$0 
(/code/ScalaMock/js/target/scala-2.12/scalamock-test-fastopt/main.js:198250)
```

After a bit of digging it turned out the JS property names are generated differently and no longer match the implementation in `org.scalamock.clazz.MockFunctionFinderImpl`.

This PR:
- (re?)enables unit tests for `shared` and `scalamockJS` - scalatest and specs were not included, so no suites were ran at all
- moves incompatible (mostly due to missing concurrency classes from stdlib or reflection usage) suites to the JVM-only module
- adds a `@JSExport` annotation to generated mock methods
- simplifies the "reflection" algorithm for SJS, which now resembles the JVM counterpart more

## Background Context

- I initially used `scalajs-stubs` to provide `JSExport` uniformly, but it didn't work on 2.11 and a check at the macro level was not too complex
- the JS property name generation algorithm is available in the [SJS emitter code](https://github.com/scala-js/scala-js/blob/a5007b08b0ba16bea2fc23e02d1767679e48b883/linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/NameGen.scala) - I've decided not to try to imitate it, as such approach turns out to be brittle

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -4,9 +4,8 @@ scalaVersion in ThisBuild := "2.11.12"
 crossScalaVersions in ThisBuild := Seq("2.11.12", "2.12.12", "2.13.3")
 //scalaJSUseRhino in ThisBuild := true
 
-lazy val scalatest = "org.scalatest" %% "scalatest" % "3.2.2"
-lazy val specs2 = "org.specs2" %% "specs2-core" % "4.10.5"
-lazy val scalameta = "org.scalameta" %% "scalameta" % "4.3.24"
+lazy val scalatest = Def.setting("org.scalatest" %%% "scalatest" % "3.2.2")
```
Comment (barkhorn) [inline build.sbt:7]: could you explain this notation to me? Is this relevant to scala.js?
the standard dependency declaration gets automatically updated by a bot, i'm not sure if it's compatible with this form

File: shared/src/main/scala/org/scalamock/util/MacroUtils.scala [hunk 2]
```diff
@@ -25,8 +25,13 @@ import MacroAdapter.Context
 /**
  * Helper functions to work with Scala macros and to create scala.reflect Trees.
  */
-private[scalamock] class MacroUtils[C <: Context](ctx: C) extends MacroAdapter[C](ctx) { // ctx2 to avoid clash with ctx in MockMaker (eugh!)
-  import ctx2.universe._
+private[scalamock] class MacroUtils[C <: Context](protected val ctx: C) extends MacroAdapter {
+  import ctx.universe._
+
+  final lazy val isScalaJs =
+    ctx.compilerSettings.exists(o => o.startsWith("-Xplugin:") && o.contains("scalajs-compiler"))
```
Comment (barkhorn) [inline shared/src/main/scala/org/scalamock/util/MacroUtils.scala:32]: I like this idea :+1:

File: js/src/main/scala/org/scalamock/clazz/MockFunctionFinderImpl.scala [hunk 3]
```diff
@@ -29,66 +29,23 @@ object MockFunctionFinderImpl {
...
-    val utils = new MacroUtils[c.type](c)
-    import utils._
-
-    def hasValueTypeArgs(baseSymbol: Symbol, owner: Type): Boolean = {
-      val baseType = owner.baseType(baseSymbol)
-      baseType.typeArgs.nonEmpty // && baseType.typeArgs.forall(_ <:< typeOf[AnyVal])
-    }
-
-    // this somehow replicates postfix logic from scala-js, there have to be a better way
-    def privateSuffix(owner: Tree): String = {
```
Comment (barkhorn) [inline js/src/main/scala/org/scalamock/clazz/MockFunctionFinderImpl.scala:41]: this has been contentious in the past, I guess newer versions of scala.js improve the toString of the `Name` type enough to make it unnecessary?
Does that introduce a dependency ont hat mechanism later on where the name gets assembled?

**Comment by barkhorn** [comment]
great PR, much appreciated! I've never worked myself with scala.js so the compatibilty has always been a bit dodgy. I've trusted CI to flag up problems, but are you saying it doesn't actually run these?
I could add an explicit goal in the script to force that then

File: js/src/main/scala/org/scalamock/clazz/MockFunctionFinderImpl.scala (hunk 3 above)
Comment (ddworak) [reply js/src/main/scala/org/scalamock/clazz/MockFunctionFinderImpl.scala:41]: With my approach I'm not depending on the name generation algorithm in Scala.js - by using `@JSExport(<encoded mock function name>)` annotation in `MockMaker` this decision is kept entirely in Scalamock. If you change `MockMakerInner#mockFunctionName`, you'll need to update this - but it's the same in the JVM impl.

File: build.sbt (hunk 1 above)
Comment (ddworak) [reply build.sbt:7]: Sure. You might already know how `%%` adds a Scala version suffix, which enables cross-building between Scala versions. The third `%` sign introduces another cross-building axis, by adding the platform part (e.g. `_sjs1`) to the artifact name, so the JS project JAR dependencies are downloaded. As can be seen in Scalamock, the JS version of the same class can differ, also the Scala.js JARs contain additional intermediate code representation and may contain JS dependencies metadata. This mechanism is actually a part of `sbt-crossproject` and works for Scala Native as well.

Due to the lack of SJS versions of ScalaTest and Specs2, no test suites were discovered in `scalamockJS/test`, so the CI was only checking whether ScalaMock compiles and links.

I believe scala-steward can handle this notation, we're using it in https://github.com/UdashFramework/udash-core where the version numbers are even harder to find, as they're extracted to a constant in another file.

**Comment by ddworak** [comment]
I'm glad you like it. I think you don't need to introduce any further steps to the CI, as `scalamock/test` always triggered `scalamockJS/test` - the problem was that there were neither ScalaTest nor specs2 JARs for Scala.js included, so no suites were discovered. You can see that the check for this PR ran `shared` suites twice, once on the JVM and once on `node`: https://travis-ci.org/github/paulbutcher/ScalaMock/jobs/744832811. That's why I had to move some of the suites to the JVM-only part, as they included reflection or concurrency mechanisms not available in the JS environment.

File: build.sbt (hunk 1 above)
Comment (barkhorn) [reply build.sbt:7]: :+1:

File: js/src/main/scala/org/scalamock/clazz/MockFunctionFinderImpl.scala (hunk 3 above)
Comment (barkhorn) [reply js/src/main/scala/org/scalamock/clazz/MockFunctionFinderImpl.scala:41]: thanks!

**Comment by barkhorn** [comment]
I've pushed v5.1.0 to maven central in case you want to update

**Pull request merged**